_posts/2017-05-11-tensorflare.md

```diff
@@ -0,0 +1,150 @@
+
+For brevity, we simply show the results of running these queries with their associated runtimes. We note, however, that Spark and PySpark have almost equivalent runtimes, with Flare and PyFlare both having approximately an 10x speedup.
+
+### Spark
+
+	scala> val q = spark.sql(tpchq6)
+    q: org.apache.spark.sql.DataFrame = [revenue: double]
+
+    scala> time(q.show)
+    +--------------------+
+    |             revenue|
+    +--------------------+
+    |1.2314107822829978E8|
+    +--------------------+
+
+    took: 11355 ms
+
+### PySpark
+
+    >>> df = spark.sql(tpchq6)
+    >>> time(df.show)
+
+	+--------------------+
+    |             revenue|
+    +--------------------+
+    |1.2314107822829978E8|
+    +--------------------+
+
+    took: 11257 ms
+
+### PyFlare
+
+	>>> df = flare(tpchq6)
+	>>> time(df.show)
+
+	+--------------------+                                      
+	|             revenue|
+	+--------------------+
+	|1.2314107822829978E8|
+	+--------------------+
+
+	took: 1217 ms
+
+### Flare
+
+	scala> val q1 = flare(q)
+    scala> time(q1.show)
+    +--------------------+    
+    |             revenue|
+    +--------------------+
+    |2.4609358141849962E8|
+    +--------------------+
+
+    took: 1161 ms
+
+## TensorFlare
+
+When looking at how to build an optimized pipeline like TensorFlare, we discovered that TensorFlow does not have a built-in mechanism for sharing tensors across sources. As such, all data querying must be followed by an extremely expensive big data copy operation. TensorFlare, however, avoids these restrictions by sharing the same intermediate representation across sources. This yields huge performance gains, without sacrificing any expressivity.
+
+To show the benefits of having such a system, we built a small experiment utilizing TensorFlow's capabilities. We first trained a model to run on US crime statistics using the method described in [TensorFlow's MNIST tutorial](https://www.tensorflow.org/get_started/mnist/beginners). We then used this model in both PySpark and TensorFlare to see what kind of results we get. In both, we run the following query:
+
+	SELECT
+		cluster,
+		sum(case when classification = 0 then 1 else 0 end) as class1, 
+		sum(case when classification = 1 then 1 else 0 end) as class2, 
+		sum(case when classification = 2 then 1 else 0 end) as class3, 
+		sum(case when classification = 3 then 1 else 0 end) as class4
+	FROM
+		(
+			SELECT 
+				cluster,
+				classifier(murder, assault, population, rape) as classification
+			FROM
+				data
+		)
+	GROUP BY cluster
+	ORDER BY cluster
+
+The classifier method mentioned above is a UDF to Spark -- a blackbox which the optimizer is unable to reason about. In our situation, this classifier is our ML model.
+
```

Comment:
I wasn't sure if you wanted me to mention the bits about sharing IR, etc., in much detail (since we just started on it)...I'll clarify that section.